packages/appium/lib/cli/setup-command.js

```diff
@@ -0,0 +1,249 @@
+import log from '../logger';
+
+/**
+ * Subcommands of preset for setup
+ */
+export const SUBCOMMAND_MOBILE = 'mobile';
+export const SUBCOMMAND_DESKTOP = 'desktop';
+export const SUBCOMMAND_BROWSER = 'browser';
+
+/**
+ * A subcommand to remove APPIUM_HOME
+ */
+export const SUBCOMMAND_RESET = 'reset';
+
+/**
+ * Pairs of preset subcommand and driver candidates.
+ * Driver names listed in KNOWN_DRIVERS to install by default
+ */
+const PRESET_PAIRS = Object.freeze(
+  /** @type {const} */ ({
+    mobile: _.keys(MOBILE_DRIVERS),
+    desktop: _.keys(DESKTOP_DRIVERS),
+    browser: _.keys(DESKTOP_BROWSERS)
+  }),
+);
+const DRIVERS_ONLY_MACOS = ['xcuitest', 'safari', 'mac2'];
+
+/**
+ * Plugin names listed in KNOWN_PLUGINS to install by default.
+ */
+export const DEFAULT_PLUGINS = ['images'];
+
+/**
+ * Return a list of drivers available for current host platform.
+ * @param {import('appium/types').CliCommandSetupSubcommand} subcmd
+ * @returns {Array<string>}
+ */
+export function getPresetDrivers(subcmd) {
+  return _.filter(PRESET_PAIRS[subcmd], (driver) => {
+    if (!system.isMac() && _.includes(DRIVERS_ONLY_MACOS, driver)) {
+      return;
+    }
+    return driver;
+  });
+}
+
+
+/**
+ * Return desktop platform name for setup command description.
+ * @returns {string}
+ */
+export function hostPlatformName() {
+  if (system.isMac()) {
+    return 'macOS';
+  } else if (system.isWindows()) {
+    return 'Windows';
+  }
+  return 'Linux';
+}
+
+/**
+ * Run 'setup' command to install drivers/plugins into the given appium home.
+ * @template {import('appium/types').CliCommandSetup} SetupCmd
+ * @param {import('appium/types').Args<SetupCmd>} preConfigArgs
+ * @param {string} appiumHome
+ * @param {DriverConfig} driverConfig
+ * @param {PluginConfig} pluginConfig
+ * @returns {Promise<void>}
+ */
+export async function runSetupCommand(appiumHome, preConfigArgs, driverConfig, pluginConfig) {
+  switch (preConfigArgs.setupCommand) {
+    case SUBCOMMAND_RESET:
+      await resetDriversPlugins(appiumHome, driverConfig, pluginConfig);
+      break;
+    case SUBCOMMAND_DESKTOP:
+      await setupDriverDesktopApp(driverConfig);
+      await setupPluginDefault(pluginConfig);
+      break;
+    case SUBCOMMAND_BROWSER:
+      await setupDriverBrowser(driverConfig);
+      await setupPluginDefault(pluginConfig);
+      break;
+    default:
+      await setupDriverMobile(driverConfig);
+      await setupPluginDefault(pluginConfig);
+      break;
+  }
+};
+
+
+/**
+ * Uninstall installed drivers and plugins in APPIUM_HOME.
+ * @param {string} appiumHome
+ * @param {DriverConfig} driverConfig
+ * @param {PluginConfig} pluginConfig
+ * @returns {Promise<void>}
+ */
+async function resetDriversPlugins(appiumHome, driverConfig, pluginConfig) {
+  log.info(`Uninstalling drivers/plugins in APPIUM_HOME[${appiumHome}]`);
+
+  for (const driverName of _.keys(driverConfig.installedExtensions)) {
+    await uninstallDriver(driverName, driverConfig);
```

Comment:
this might fail if the config is corrupted. That is why it makes sense to enclose each uninstall call into try/catch and delete the config yaml completely in the end.